test/CMakeLists.txt

```diff
@@ -0,0 +1,433 @@
+    return()
+  endif()
+
+#>>emily
+# file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/${FILE_COLLECTION} DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+#<<emily
+
+  add_custom_target(get_obsforge_test_data ALL DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${FILE_COLLECTION})
+  add_custom_command(
+    TARGET get_obsforge_test_data
+    POST_BUILD
+    COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_BINARY_DIR} tar xzf ${FILE_COLLECTION}
+    COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_BINARY_DIR} rm -rf testdata
+    COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_BINARY_DIR} rm -rf testoutput
+    COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_BINARY_DIR} mv remote_data/testdata testdata
+    COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_BINARY_DIR} mv remote_data/testoutput testoutput
+    COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_BINARY_DIR} rm -rf remote_data)
+endif()
+
+message("Emily checking test/CMAKE_CURRENT_SOURCE_DIR - ${CMAKE_CURRENT_SOURCE_DIR}")
+message("Emily checking test/CMAKE_CURRENT_BINARY_DIR - ${CMAKE_CURRENT_BINARY_DIR}")
+
+# Function definitions
+
+# The list of all test input config files, which need to be symlink from the build folder.
+# (These symlink are created at the end of this file, to give CMake functions setting up certain
+# tests a chance to append items to this list).
+
+# List of test cases
+list( APPEND test_case
+  bufr2netcdf
+  script2netcdf
+  bufr4backend 
+  script4backend 
+)
+
+# List of test input files 
+list( APPEND test_input
+  bufr_satwnd_amv_goes_mapping.yaml
+  bufr_satwnd_amv_goes.py
+  bufr_sfcsno_mapping.yaml
+  bufr_sfcsno.py
+)
+
+# List of test configurations 
+list( APPEND test_config
+  bufr_bufr4backend_satwnd_amv_goes.yaml
+  bufr_bufr4backend_satwnd_amv_goes_mpi4.yaml
+  bufr_script4backend_satwnd_amv_goes.yaml
+  bufr_script4backend_satwnd_amv_goes_mpi4.yaml
+  bufr_bufr4backend_sfcsno.yaml
+  bufr_bufr4backend_sfcsno_mpi2.yaml
+  bufr_script4backend_sfcsno.yaml
+  bufr_script4backend_sfcsno_mpi2.yaml
+)
+
+# create test intput directories and make links to the input files
+file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/testinput)
+foreach(FILENAME ${test_input})
+  execute_process( COMMAND ${CMAKE_COMMAND} -E create_symlink
+    ${CMAKE_CURRENT_SOURCE_DIR}/../sorc/spoc/dump/mapping/${FILENAME}
+    ${CMAKE_CURRENT_BINARY_DIR}/testinput/${FILENAME} )
+endforeach(FILENAME)
+
+foreach(FILENAME ${test_config})
+  execute_process( COMMAND ${CMAKE_COMMAND} -E create_symlink
+    ${CMAKE_CURRENT_SOURCE_DIR}/testinput/${FILENAME}
+    ${CMAKE_CURRENT_BINARY_DIR}/testinput/${FILENAME} )
+endforeach(FILENAME)
+
+# create test output directories
+foreach(TESTCASE ${test_case})
+  file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/testrun/${TESTCASE})
+endforeach(TESTCASE)
+
+# =====================================================================
+# Tests for sfcsno 
+# =====================================================================
+# Perform bufr2netcdf tests
```

Comment:
for future (not in this PR) can we create functions so that if we have `n` number of tests for each ob type `i`, we can create `n` functions to generate `n*i` tests?

Reply:
I am not sure how to do this yet.  But, I will find out how.